[Merged by Bors] - feat(LinearAlgebra/Matrix/PosDef): `vecMulVec a (star a)` is positive semi-definite

themathqueen opened the pull request:
The matrices `vecMulVec a (star a)` and `vecMulVec (star a) a` are always positive semi-definite.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

themathqueen committed:
easy lemmas

eric-wieser reviewed (commented):
Is there a `PosDef` version of this too?

themathqueen committed:
posDef versions

themathqueen committed:
reorder

themathqueen commented:
> Is there a `PosDef` version of this too?

Ye sure. It would need `(replicateRow Unit a).mulVec` (and `replicateCol` respectively) to be injective I think.

eric-wieser commented on Mathlib/LinearAlgebra/Matrix/PosDef.lean:
```diff
@@ -558,6 +566,24 @@ theorem conjTranspose_mul_self [StarOrderedRing R] [NoZeroDivisors R] (A : Matri
+  exact mul_conjTranspose_self _ ha
+
+theorem _root_.Matrix.posDef_star_vecMulVec [StarOrderedRing R] [NoZeroDivisors R] (a : n → R)
+    (ha : Function.Injective (replicateRow Unit a).mulVec) :
```
```suggestion
    (ha : Function.Injective (dotProduct a) :
```

themathqueen committed:
dotProduct instead

themathqueen commented on Mathlib/LinearAlgebra/Matrix/PosDef.lean:
```diff
@@ -558,6 +566,24 @@ theorem conjTranspose_mul_self [StarOrderedRing R] [NoZeroDivisors R] (A : Matri
+  simpa using mul_mul_conjTranspose_same .one hA
+
+theorem _root_.Matrix.posDef_vecMulVec_star [StarOrderedRing R] [NoZeroDivisors R] (a : n → R)
+    (ha : Function.Injective (replicateCol Unit a).vecMul) :
```
I'll need to think about it 🤔

themathqueen replied:
@eric-wieser, should I commit the above? If so, where would I put the first one?

themathqueen committed:
fix

themathqueen committed:
vecMulVec_star not posDef

themathqueen committed:
docstrings

themathqueen committed:
rename to match posSemidef

themathqueen committed:
Merge branch 'master' into posSemidef_vecMulVec

ocfnash reviewed (commented):
Thanks for all of this.

The code is great but I'm a bit dubious about some of the statements. Maybe I can be persuaded :) I've left some comments in line!

themathqueen committed:
Update Mathlib/Data/Matrix/Mul.lean

themathqueen committed:
Update Mathlib/LinearAlgebra/Matrix/PosDef.lean

ocfnash reviewed (approved):
Thanks for this!

Please drop the lemma `Matrix.not_posDef_vecMulVec` and then feel free to merge.

bors d+

themathqueen committed:
Remove theorem on vecMulVec non-positivity

Removed theorem stating that certain vector products are not positive definite in nontrivial rings.
